### PR TITLE
CORE-38: Merge dev to master for git reorganization.

### PR DESCRIPTION
**Pull request by ghost**
Bringing master up to date with current development.

### CODE REVIEW

**Comment by kirkwood** [comment]
### Note on comms subsystem in this pull request
@leithaus 
The diff is hard to wrap one's head around, but this PR simply cleans out _everything_ that was in the `comm/` subdirectory in `master` and replaces it with the bare minimum:
* The Kademlia table implementation [here](https://github.com/rchain/rchain/tree/dev/comm/src/main/scala/coop/rchain/kademlia) and its unit tests [here](https://github.com/rchain/rchain/tree/dev/comm/src/test/scala/coop/rchain/kademlia)
* The pure Java zeromq implementation [here](https://github.com/rchain/rchain/tree/dev/comm/src/main/java)
* Protocol buffer stubs for peer-to-peer routing [here](https://github.com/rchain/rchain/tree/dev/comm/src/main/protobuf/coop/rchain/comm/protocol)

Since most of the code I've produced hasn't been reviewed, I will reintroduce it to `dev` in smaller chunks as it gets reworked.

File: comm/src/main/scala/coop/rchain/comm/network.scala [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        comm.recv match {
+          case Success(res) =>
+            for {
+              msg <- ProtocolMessage.parse(res)
+            } dispatch(msg)
+          case Failure(ex: SocketTimeoutException) => ()
+          case Failure(ex)                         => ex.printStackTrace
+        }
+      }
+  }.start
```
Comment (DoSOfRedRiver) [inline comm/src/main/scala/coop/rchain/comm/network.scala]: Do you really need the `receiver` variable? The `Thread#start` method returns `Unit`.
Also, I found another way to write this code. Scala 2.12 has SAM support, so you can write something like this:
```val task: Runnable = () => { //body }
val receiver = new Thread(task).start()```

**Comment by kirkwood** [comment]
I can't find this comment in github.

You're right, though. I (eventually) do need to remember the thread so it
can be stopped; there's currently no way to do that. I have a future PR
with that in it.

Chris.

**Comment by pyrocto** [comment]
Since we decided to involve @leithaus not in every push to master but rather when we cut releases, I'm going to merge this now.

**Pull request merged**